### PR TITLE
allow setting sharedContainerIdentifier to allow background url sessi…

### PR DESCRIPTION
**Pull request by thuyBox**
…on working in app extension and add sample app's share extension allowing upload image file

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @thuyBox has signed the CLA. Thanks for the pull request!

File: BoxContentSDKSampleApp/Podfile.lock [hunk 1]
```diff
@@ -14,8 +14,6 @@ EXTERNAL SOURCES:
     :path: "../"
 
 SPEC CHECKSUMS:
-  box-ios-sdk: 8f8663807e26fd5e6bb89a07c0402d2bef605d30
-
-PODFILE CHECKSUM: 665c09eb2d2e131c961508f8a52594e06e23dab9
```
Comment (adempsey) [inline BoxContentSDKSampleApp/Podfile.lock]: do we not need this?

File: BoxContentSDKSampleApp/shareExtension/Info.plist [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	<key>CFBundleShortVersionString</key>
+	<string>1.0</string>
+	<key>CFBundleVersion</key>
+	<string>1</string>
+	<key>NSExtension</key>
+	<dict>
+		<key>NSExtensionAttributes</key>
+		<dict>
+			<key>NSExtensionActivationRule</key>
+			<string>TRUEPREDICATE</string>
```
Comment (adempsey) [inline BoxContentSDKSampleApp/shareExtension/Info.plist]: the code in `ShareViewController` appears to assume only images are being uploaded. In that case, we should modify the activation rule to align with that.

File: BoxContentSDKSampleApp/shareExtension/ShareViewController.m [hunk 3]
```diff
@@ -0,0 +1,80 @@
+//
+//  ShareViewController.m
+//  shareExtension
+//
+//  Created by Thuy Nguyen on 5/8/17.
+//  Copyright © 2017 Box. All rights reserved.
+//
+
+#import "ShareViewController.h"
+#import <BoxContentSDK/BOXContentSDK.h>
```
Comment (adempsey) [inline BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: `@import BOXContentSDK;`

File: BoxContentSDKSampleApp/shareExtension/ShareViewController.m [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+- (void)upload:(NSURL *)url
+{
+    [self setUp];
+    NSString *path = url.path;
+    NSString *tempPath = [path stringByAppendingString:@".temp"];
+    NSString *associateId = [BOXSampleAppSessionManager generateRandomStringWithLength:32];
+
+    BOXFileUploadRequest *uploadRequest = [self.client fileUploadRequestInBackgroundToFolderWithID:@"6598717269" fromLocalFilePath:path uploadMultipartCopyFilePath:tempPath associateId:associateId];
```
Comment (adempsey) [inline BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: What's this folder ID? Should we be using root?

File: BoxContentSDKSampleApp/shareExtension/ShareViewController.m [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    uploadRequest.enableCheckForCorruptionInTransit = YES;
+
+    [uploadRequest performRequestWithProgress:^(long long totalBytesTransferred, long long totalBytesExpectedToTransfer) {
+        NSLog(@"----->>>>%lld, %lld", totalBytesTransferred, totalBytesExpectedToTransfer);
+    } completion:^(BOXFile *file, NSError *error) {
+        NSLog(@"----->>>>file %@, error %@", file, error);
```
Comment (adempsey) [inline BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: We should use `BOXLog` and avoid the `---->>>>` prefix

File: BoxContentSDKSampleApp/shareExtension/ShareViewController.m [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    }];
+}
+
+- (void)didSelectPost {
+    NSString *typeIdentifier = @"public.image";
```
Comment (adempsey) [inline BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: should use the `kUTTypeImage` constant

File: BoxContentSDKSampleApp/shareExtension/ShareViewController.m [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+@implementation ShareViewController
+
+- (void)setUp
+{
+#warning Set the client ID and client secret that can be retrieved by creating an application at http://developers.box.com
+
+    [BOXContentClient setClientID:@"your_client_id" clientSecret:@"your_client_secret"];
+
+    [BOXContentClient oneTimeSetUpInExtensionToSupportBackgroundTasksWithDelegate:self rootCacheDir:[BOXSampleAppSessionManager rootCacheDir] sharedContainerIdentifier:@"group.com.box.ios.dev" completion:^(NSError *error) {
```
Comment (adempsey) [inline BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: This is the same container identifier that the Box app uses. Should the sample app use a different group?

File: BoxContentSDKSampleApp/shareExtension/ShareViewController.m [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+@property (nonatomic, strong, readwrite) BOXContentClient *client;
+
+@end
+
+@implementation ShareViewController
+
+- (void)setUp
+{
+#warning Set the client ID and client secret that can be retrieved by creating an application at http://developers.box.com
```
Comment (adempsey) [inline BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: maybe even `#error` - not much use to running this code if these aren't set

**Review by adempsey — COMMENTED** [review]
Does this work with large files? iOS doesn't terminate the extension while the upload is still occurring?

File: BoxContentSDKSampleApp/shareExtension/Info.plist (hunk 2 above)
Comment (thuyBox) [reply BoxContentSDKSampleApp/shareExtension/Info.plist]: done

File: BoxContentSDKSampleApp/shareExtension/ShareViewController.m (hunk 7 above)
Comment (thuyBox) [reply BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: updated to different container

File: BoxContentSDKSampleApp/shareExtension/ShareViewController.m (hunk 4 above)
Comment (thuyBox) [reply BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: updated to root

File: BoxContentSDKSampleApp/shareExtension/ShareViewController.m (hunk 5 above)
Comment (thuyBox) [reply BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: done

File: BoxContentSDKSampleApp/shareExtension/ShareViewController.m (hunk 6 above)
Comment (thuyBox) [reply BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: done

File: BoxContentSDKSampleApp/shareExtension/ShareViewController.m (hunk 8 above)
Comment (thuyBox) [reply BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: done

File: BoxContentSDKSampleApp/Podfile.lock (hunk 1 above)
Comment (thuyBox) [reply BoxContentSDKSampleApp/Podfile.lock]: done

**Comment by thuyBox** [comment]
@adempsey the upload will still run in the background even if the extension terminates

File: BoxContentSDKSampleApp/shareExtension/ShareViewController.m [hunk 9]
```diff
@@ -46,18 +46,18 @@ - (void)upload:(NSURL *)url
...
+        BOXLog(@"totalBytesTransferred, totalBytesExpectedToTransfer: %lld, %lld", totalBytesTransferred, totalBytesExpectedToTransfer);
     } completion:^(BOXFile *file, NSError *error) {
-        NSLog(@"----->>>>file %@, error %@", file, error);
+        BOXLog(@"upload request finished with file %@, error %@", file, error);
     }];
 }
 
 - (void)didSelectPost {
-    NSString *typeIdentifier = @"public.image";
+    NSString *typeIdentifier = (NSString *)kUTTypeImage;
```
Comment (adempsey) [inline BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: I think this needs to be `(__bridge_transfer NSString *)`

File: BoxContentSDKSampleApp/shareExtension/ShareViewController.m (hunk 3 above)
Comment (adempsey) [reply BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: ^ Still this one

File: BoxContentSDKSampleApp/Podfile.lock [hunk 10]
```diff
@@ -16,4 +16,6 @@ EXTERNAL SOURCES:
 SPEC CHECKSUMS:
   box-ios-sdk: 5a7ff7f3c0604db1bf5c8a87e090c9711b007afe
 
-COCOAPODS: 1.2.0
+PODFILE CHECKSUM: 55d216b7cce66dd8cc47218d16176fad94884b41
+
+COCOAPODS: 1.2.1
```
Comment (adempsey) [inline BoxContentSDKSampleApp/Podfile.lock]: do we need to specify `1.2.1`?

File: BoxContentSDKSampleApp/shareExtension/ShareViewController.m [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+- (BOOL)isContentValid {
+    // Do validation of contentText and/or NSExtensionContext attachments here
+    return YES;
+}
+
+- (void)upload:(NSURL *)url
+{
+    [self setUp];
+    NSString *path = url.path;
+    NSString *tempPath = [path stringByAppendingString:@".temp"];
```
Comment (adempsey) [inline BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: should use `stringByAppendingPathExtension`

Comment (thuyBox) [reply BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: done

File: BoxContentSDKSampleApp/shareExtension/ShareViewController.m (hunk 9 above)
Comment (thuyBox) [reply BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: done

File: BoxContentSDKSampleApp/Podfile.lock (hunk 10 above)
Comment (thuyBox) [reply BoxContentSDKSampleApp/Podfile.lock]: we don't. updated

File: BoxContentSDKSampleApp/shareExtension/ShareViewController.m [hunk 12]
```diff
@@ -43,7 +43,7 @@ - (void)upload:(NSURL *)url
 {
     [self setUp];
     NSString *path = url.path;
-    NSString *tempPath = [path stringByAppendingString:@".temp"];
+    NSString *tempPath = [path stringByAppendingPathExtension:@".temp"];
```
Comment (adempsey) [inline BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: should append just `@"temp"`

Comment (thuyBox) [reply BoxContentSDKSampleApp/shareExtension/ShareViewController.m]: done

**Review by adempsey — APPROVED** [review]

**Pull request merged**